langchain/chains/chat_vector_db/base.py

```diff
@@ -83,3 +84,22 @@ def _call(self, inputs: Dict[str, Any]) -> Dict[str, str]:
         new_inputs["chat_history"] = chat_history_str
         answer, _ = self.combine_docs_chain.combine_docs(docs, **new_inputs)
         return {self.output_key: answer}
+
+    async def _acall(self, inputs: Dict[str, Any]) -> Dict[str, str]:
+        question = inputs["question"]
+        chat_history_str = _get_chat_history(inputs["chat_history"])
+        if chat_history_str:
+            new_question = await self.question_generator.arun(
+                question=question, chat_history=chat_history_str
+            )
+        else:
+            new_question = question
+        # Run similarity search in executor to avoid blocking event loop
+        docs = await asyncio.get_event_loop().run_in_executor(
+            None, self.vectorstore.similarity_search, new_question, 4
```

Comment:
@nfcampos - It's a hassle to async-ify all vector store implementations so for now I'm running them in threadpool executor to avoid blocking main runloop. However, since vectorstore implementations are not guaranteed to be threadsafe, and with the default executor, you can have multiple threads, I'm not sure this is the best approach. Thoughts here?

Reply:
Hmm that's an interesting problem. The fundamental issue I think (but correct me if I'm wrong) is that, for maximum throughput, we want eg multiple simultaneous calls into this chain to each call into the vector store and 1. not block other things in the event loop and 2. not get blocked waiting on other vector store calls?

So if we somehow placed the vector store into its own dedicated thread (or something to that effect) that wouldn be enough to solve (1) but not (2) as multiple simultaneous calls to it would have to be queued up?

Looking at each vector store, the ones that do just do http requests I assume could be made async or threadsafe sync trivially. The ones that could be non-trivially not-thread safe are the ones that run locally, so faiss and chroma? Faiss looks to be thread safe for reads (not writes) https://github.com/facebookresearch/faiss/wiki/Threads-and-asynchronous-calls, don't know about chroma

Thoughts?

Reply:
Yeah those are great points. Not many people are using the async api rn and getting this in is blocking some stuff we want to release tmrw so we'll probably merge by taking out the "run_in_executor" call to avoid any potential wonky behavior from thread interactions.